Luxor/vested wg spending

dobertRowneySr opened the pull request:
Addresses: #4948 
based on top of #5062 

Note: I have used the extrinsic `vested_transfer` inside the extrinsic `spend_from_budget` all extrinsic in substrate are now `transactional` by default, this means that if the `vested_transfer` fails then `spend_from_budget` is equivalent to a no-op.
Also I have created one internal account for each WG where tokens are minted before they are transferred with vesting.

Commit:
feat: :art: add extrinsic for decreasing budgetj

Commit:
test: :test_tube: add tests for new extrinsic

Commit:
feat: :art: add proposal and tests to proposal pallet

Commit:
feat: :art: add proposal config to runtime

Commit:
fix: :zap: rearrange checks

Commit:
test: :zap: add proposal to integration test and update types and metadatat

Commit:
feat: :art: metadata 2003

Commit:
Revert "test: :zap: add proposal to integration test and update types and metadatat"

This reverts commit 4041f80159c7bbc93aeac1657de8898e57a1d685.

Commit:
feat: :zap: fix error on integration tests code

Commit:
feat: :art: add functionality for vested budget spending

Commit:
fix: :green_heart: build errors and add tests

Commit:
fix: :green_heart: lookup argument

Commit:
fix: :green_heart: add converter in order to silence compiler

Commit:
test: :white_check_mark: add test on vesting

Commit:
feat: :art: fix integartion tests

Commit:
fix: :white_check_mark: test setup CI build

Commit:
fix: :white_check_mark: fix mock setup for test on storage pallet

Commit:
fix: :bug: issue with amm sale

Commit:
fix: :green_heart: fix ci checks

Commit:
fix: :green_heart: checks for benchmarks

Commit:
fix: :white_check_mark: benchmarks

dobertRowneySr commented:
To be rebased onto #5062

Commit:
feat: :sparkles: rebase on latest luxor

Commit:
feat: :sparkles: update with latest luxor

mnaamani commented:
I think there is a simpler way to do this, without needing to have an account on each working group.
Instead of `pay_from_budget` into the working group account then doing `vesting::vested_transfer` you should be able to pay from the budget directly into the target account, then do `vesting::force_vested_transfer` with source and target being the final destination. The balances::transfer inside force_vested_transfer will be a successful no-op, and then the vesting schedule will be applied.

mnaamani commented:
> Also I have created one internal account for each WG where tokens are minted before they are transferred with vesting.

https://github.com/Joystream/joystream/pull/5072#issuecomment-2002164477

mnaamani commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1140,31 +1152,60 @@ decl_module! {
+                }
+
+                // first mint into the treasury account
+                Self::pay_from_budget(&treasury_account_id, amount);
```
ref: https://github.com/Joystream/joystream/pull/5072#issuecomment-2002164477

```rust
Self::pay_from_budget(account_id, amount);
vesting::Pallet::<T>::force_vested_transfer(
  RawOrigin::Root,
  account_id,
  account_id,
  vesting_schedule
)
```

mnaamani reviewed (changes requested):
Can you confirm that [force_vested_transfer](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/vesting/src/lib.rs#L371) the way I'm suggested can be done to simplify this implementation?

Commit:
feat: :art: simplify code with `force_vested_transfer`

The amount is first transferred in full to the worker reward account and immediately after lock is applied

Commit:
Merge branch 'luxor' into pr/ignazio-bovo/5072

Commit:
fix: :zap: update with latest luxor

dobertRowneySr commented:
I need to add benchmarks

mnaamani commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1626,6 +1681,11 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
     }
+
+    /// Returns the account for the  module used for vested spending
+    pub fn module_treasury_account() -> T::AccountId {
```
this can be dropped

mnaamani commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -129,6 +137,9 @@ pub trait Config<I: Instance = DefaultInstance>:
     type LeaderOpeningStake: Get<Self::Balance>;
+
+    /// Type for the Module Id account in order to locked payment the budget.
+    type ModuleId: Get<PalletId>;
```
As we no longer need to store funds in the working group, this can be dropped

mnaamani commented on runtime-modules/working-group/src/tests/fixtures.rs:
```diff
@@ -1159,6 +1162,78 @@ impl SetStatusTextFixture {
+
+        if actual_result.is_ok() {
+            assert_eq!(new_budget, old_budget - amount);
+            assert_eq!(new_balance, new_balance);
```
not sure what is being checked here?

dobertRowneySr replied:
legacy code that I did not remove..

mnaamani commented:
I think this now needs to be rebased on luxor since I merged the reduce council budget PR

Commit:
feat: :art: add benchmarks

Commit:
Merge branch 'luxor' into pr/ignazio-bovo/5072

Commit:
fix: :memo: address pr comments

thesan commented on tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts:
```diff
@@ -9,15 +9,27 @@ import { ISubmittableResult } from '@polkadot/types/types/'
 export class SpendBudgetFixture extends BaseWorkingGroupFixture {
   protected recievers: string[]
-  protected amounts: BN[]
+  protected vestingSchedules: VestingSchedule[]
```
Should this be changed ? The fixture is still called `SpendBudgetFixture` not `SpendVestedBudgetFixture`

dobertRowneySr replied:
change reverted, integration tests needs to be taken care of in a separate PR. 
Namely add this Fixture with Flow

Commit:
fix: :zap: integrations tests

Commit:
fix: :bug: revert metadata 2002.json to master version and fix issue on network tests

mnaamani commented on runtime-modules/content/src/tests/mock.rs:
```diff
@@ -476,6 +479,15 @@ impl common::council::CouncilBudgetManager<U256, u64> for CouncilBudgetManager {
+parameter_types! {
+    pub const DWGModuleId: PalletId = PalletId(*b"m_distwg");
+    pub const SWGModuleId: PalletId = PalletId(*b"m_storwg");
+    pub const CWGModuleId: PalletId = PalletId(*b"m_contwg");
```
ModuleId not required anymore

mnaamani commented on runtime-modules/content/src/tests/mock.rs:
```diff
@@ -490,6 +502,8 @@ impl working_group::Config<StorageWorkingGroupInstance> for Test {
     type MinimumApplicationStake = MinimumApplicationStake;
     type LeaderOpeningStake = LeaderOpeningStake;
+    type VestingBalanceToBalance = BalanceConverter;
+    type ModuleId = SWGModuleId;
```
ModuleId not a configuration trait anymore so should be removed

mnaamani commented on runtime/src/lib.rs:
```diff
@@ -1355,6 +1355,7 @@ parameter_types! {
     pub const MinimumApplicationStake: Balance = dollars!(20);
     // This should be more costly than `add_opening` fee
     pub const LeaderOpeningStake: Balance = dollars!(100);
+    pub const WorkingGroupModuleId: PalletId = PalletId(*b"mworking"); // module storage
```
not required

mnaamani commented on tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts:
```diff
@@ -9,6 +9,12 @@ import { ISubmittableResult } from '@polkadot/types/types/'
 import { Utils } from '../../utils'
 import { BudgetSpendingEventFieldsFragment, WorkingGroupFieldsFragment } from '../../graphql/generated/queries'
 
+export type VestingSchedule = {
```
unused type, I think you mentioned integration tests will be updated in a separate PR?

dobertRowneySr replied:
yes

mnaamani reviewed (changes requested):
some additional cleanup for build to works after refactor (removing the moduleId in working groups)

Commit:
fix: :test_tube: tests & mocks

Commit:
fix: :test_tube: benchmark error on amount type

mnaamani reviewed (commented):
Benchmark is failing

```
Error: Input("Benchmark working_group::vested_spend_from_budget failed: AmountLow")
2024-03-29 22:54:25 Starting benchmark: working_group::vested_spend_from_budget    
There was a problem generating the weights for working_group, check the error above
Error: Process completed with exit code 1.
```

mnaamani commented on runtime-modules/working-group/src/benchmarking.rs:
```diff
@@ -787,12 +788,34 @@ benchmarks_instance! {
+            None
+        );
+
+        let amount_vesting = VestingBalanceOf::<T>::from(10_000u32);
```
Benchmark is failing with error here`AmountLow` because amount is less than vesting pallet configuration of `<T as vesting::Config>::MinVestedTransfer::get()`

mnaamani replied:
So perhaps use that constant to make the `amount_vesting` at least that amount.

mnaamani commented on runtime-modules/working-group/src/benchmarking.rs:
```diff
@@ -787,12 +788,34 @@ benchmarks_instance! {
+
+        let amount_vesting = VestingBalanceOf::<T>::from(10_000u32);
+        let budget = BalanceOf::<T>::from(100_000u32);
+        let block_no = 100u32;
```
`block_no` is not used

Commit:
fix: :green_heart: benchmarks

Commit:
fix: :zap: benchmarking

mnaamani reviewed (changes requested):
from earlier

Commit:
fix: :zap: benchmarks

Commit:
fix: :green_heart: build errors

Commit:
fix: :zap: runtime test errors

Commit:
fix: :green_heart: ci errors

Commit:
fix: :green_heart: round of ci fixes

mnaamani commented:
strange that I'm seeing errors like `../../query-node/start.sh: line 11: docker-compose: command not found` in CI workflow runs..
Has it been deprecated and we should replace all occurrences of that command with `docker compose` command?

I think you forgot to commit changes to: `types/src/augment/augment-api-rpc.ts` and  `types/src/augment/augment-types.ts`

mnaamani reviewed (approved):
LGTM, next PR to luxor should commit changes to types/src/*.rs missed in this PR

Running integration tests locally on my machine works, we need to investigate why they are failing on github runners.
